src/Pos/Client/Txp/History.hs

```diff
@@ -102,86 +90,106 @@ data TxHistoryEntry = THEntry
 
 -- | Given a full blockchain, derive address history and Utxo
 -- TODO: Such functionality will still be useful for merging
 -- blockchains when wallet state is ready, but some metadata for
 -- Tx will be required.
 deriveAddrHistory
-    -- :: (Monad m, Ssc ssc) => Address -> [Block ssc] -> TxSelectorT m [TxHistoryEntry]
-    :: (Monad m) => [Address] -> [Block ssc] -> TxSelectorT m [TxHistoryEntry]
-deriveAddrHistory addr chain = do
-    ether $ identity %= filterUtxoByAddrs addr
-    deriveAddrHistoryPartial [] addr chain
-
-deriveAddrHistoryPartial
-    :: (Monad m)
-    => [TxHistoryEntry]
-    -> [Address]
-    -> [Block ssc]
-    -> TxSelectorT m [TxHistoryEntry]
-deriveAddrHistoryPartial hist addrs chain = deriveAddrHistoryPartialWithTimestamp hist addrs chain Nothing
-
-deriveAddrHistoryPartialWithTimestamp
-    :: (Monad m)
-    => [TxHistoryEntry]
-    -> [Address]
-    -> [Block ssc]
-    -> Maybe SlottingData
-    -> TxSelectorT m [TxHistoryEntry]
-deriveAddrHistoryPartialWithTimestamp hist addrs chain maybeSd =
-    DL.toList <$> foldrM updateAll (DL.fromList hist) chain
-  where
-    updateAll (Left _) hst = pure hst
-    updateAll (Right blk) hst = do
-        let mapper TxAux {..} = (withHash taTx, taWitness, taDistribution)
-        txs <- getRelatedTxs addrs $
-                   map mapper $ flattenTxPayload (blk ^. mainBlockTxPayload)
-        let difficulty = blk ^. difficultyL
-            txs' = map (thDifficulty .~ Just difficulty) txs
-        let maybeTimestamp = maybeSd >>= getSlotStartPure True (blk ^. mainBlockSlot)
-            txs'' = map (thTimestamp .~ maybeTimestamp) txs'
-        return $ DL.fromList txs'' <> hst
+    :: MonadUtxo m
+    => [Address] -> [Block ssc] -> m [TxHistoryEntry]
+deriveAddrHistory addrs chain =
+    DL.toList <$> foldrM (flip $ deriveAddrHistoryBlk addrs $ const Nothing) mempty chain
+
+deriveAddrHistoryBlk
+    :: MonadUtxo m
+    => [Address]
+    -> (MainBlock ssc -> Maybe Timestamp)
+    -> DList TxHistoryEntry
+    -> Block ssc
+    -> m (DList TxHistoryEntry)
+deriveAddrHistoryBlk _ _ hist (Left _) = pure hist
+deriveAddrHistoryBlk addrs getTs hist (Right blk) = do
+    let mapper TxAux {..} = (withHash taTx, taWitness, taDistribution)
+        difficulty = blk ^. difficultyL
+        mTimestamp = getTs blk
+    txs <- getRelatedTxsByAddrs addrs (Just difficulty) mTimestamp $
+           map mapper . flattenTxPayload $
+           blk ^. mainBlockTxPayload
+    return $ DL.fromList txs <> hist
+
+----------------------------------------------------------------------------
+-- GenesisToil
+----------------------------------------------------------------------------
+
+-- | Identity wrapper to use genesis utxo in context as `MonadUtxoRead` instance
+-- TODO: probably should be moved elsewhere; `Pos.Txp.Toil` is not possible, because
+-- of dependency on `Pos.Context` from main package
+data GenesisToilTag
+
+type GenesisToil = Ether.TaggedTrans GenesisToilTag IdentityT
```

Comment:
minor:
you can use `UtxoReaderT`, like here
https://github.com/input-output-hk/cardano-sl/blob/cardano-sl-0.5-rc/src/Pos/Wallet/Web/Tracking.hs#L267